### PR TITLE
FOUNDATIONS: Implement DB Context Base

### LINKED ISSUE #18: FOUNDATIONS: Implement DB Context Base

### PR DESCRIPTION
**Pull request by ZafarUrakov**
Closes #18 
@cjdutoit 
@ElbekDeveloper

### CODE REVIEW

File: EFxceptions.Core/DBContextBase.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            InitializeInternalServices();
+
+        private void InitializeInternalServices()
+        {
+            this.cosmosErrorBroker = CreateErrorBroker();
+            this.eFxceptionService = CreateEFxceptionService(this.cosmosErrorBroker);
+        }
+
+        protected abstract ICosmosErrorBroker CreateErrorBroker();
+        protected abstract IEFxceptionService CreateEFxceptionService(
```
Comment (cjdutoit) [inline EFxceptions.Core/DBContextBase.cs:32]: Make a one lainer or add a new line above

File: EFxceptions.Core/DBContextBase.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            try
+            {
+                return await base.SaveChangesAsync(cancellationToken);
+            }
+            catch (CosmosException cosmosException)
+            {
+                this.eFxceptionService.ThrowMeaningfulException(
+                    cosmosException);
+
+                throw;
```
Comment (cjdutoit) [inline EFxceptions.Core/DBContextBase.cs:47]: Unreached code.   remove `throw;` here and and the other methods

File: EFxceptions.Core/DBContextBase.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        {
+            try
+            {
+                return await base.SaveChangesAsync(acceptAllChangesOnSuccess, cancellationToken);
+            }
+            catch (CosmosException cosmosException)
+            {
+                this.eFxceptionService.ThrowMeaningfulException(cosmosException);
+
+                throw;
```
Comment (cjdutoit) [inline EFxceptions.Core/DBContextBase.cs:63]:
```suggestion
                this.eFxceptionService.ThrowMeaningfulException(cosmosException);
                throw;
```

File: EFxceptions.Core/DBContextBase.cs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        {
+            try
+            {
+                return base.SaveChanges();
+            }
+            catch (CosmosException cosmosException)
+            {
+                this.eFxceptionService.ThrowMeaningfulException(cosmosException);
+
+                throw;
```
Comment (cjdutoit) [inline EFxceptions.Core/DBContextBase.cs:77]:
```suggestion
                this.eFxceptionService.ThrowMeaningfulException(cosmosException);
                throw;
```

File: EFxceptions.Core/DBContextBase.cs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        {
+            try
+            {
+                return base.SaveChanges(acceptAllChangesOnSuccess);
+            }
+            catch (CosmosException cosmosException)
+            {
+                this.eFxceptionService.ThrowMeaningfulException(cosmosException);
+
+                throw;
```
Comment (cjdutoit) [inline EFxceptions.Core/DBContextBase.cs:92]:
```suggestion
                this.eFxceptionService.ThrowMeaningfulException(cosmosException);
                throw;
```

File: EFxceptions.Core/DBContextBase.cs (hunk 3 above)
Comment (ZafarUrakov) [reply EFxceptions.Core/DBContextBase.cs]: Done

**Review by cjdutoit — APPROVED** [review]

**Pull request merged**